miden-lib/asm/kernels/transaction/api.masm

```diff
@@ -863,20 +864,155 @@ export.get_note_serial_number
+#!     CODE_ROOT: [[ACCOUNT_PROCEDURE_DATA]]
+#!   }
+#! Outputs:
+#!   Operand stack: []
 #!
 #! Where:
-#! - foreign_account_id is the ID of the foreign account whose procedure is going to be executed. 
+#! - foreign_account_id is the ID of the foreign account whose procedure is going to be executed.
+#! - FOREIGN_ACCOUNT_ID is the word constructed from the foreign_account_id as follows:
+#!   [foreign_account_id, 0, 0, 0].
+#! - account_nonce is the nonce of the foreign account.
+#! - VAULT_ROOT is the commitment of the foreign account's vault.
+#! - STORAGE_ROOT is the commitment of the foreign account's storage.
+#! - STORAGE_SLOT_DATA is the data contained in the storage slot which is constructed as follows: 
+#!   [SLOT_VALUE, slot_type, 0, 0, 0]
+#! - CODE_ROOT is the commitment of the foreign account's code.
+#! - ACCOUNT_PROCEDURE_DATA is the information about account procedure which is constructed as 
+#!   follows: [PROCEDURE_MAST_ROOT, storage_offset, 0, 0, 0]
 #!
 #! Panics if:
 #! - the current context is not a native context.
-export.start_foreign_context
+export.start_foreign_context.1
     # check that this procedure was executed against the native account
     exec.memory::assert_native_account
+    # OS => [KERNEL_PROCEDURE_HASH, foreign_account_id]
 
-    dropw drop # unimplemented!
+    # drop the procedure's hash
+    dropw
+    # OS => [foreign_account_id]
+
+    # get the memory address and a flag whether this account was already loaded.
+    exec.memory::get_foreign_account_ptr
+    # OS => [was_loaded, ptr, foreign_account_id]
+
+    if.true
+        exec.memory::set_current_account_data_ptr drop
+        # OS => []
+    else
+        exec.memory::set_current_account_data_ptr
+        # OS => [foreign_account_id]
+
+        # construct the word with account ID to load the core account data from the advice map
+        push.0.0.0 movup.3
+        # OS => [foreign_account_id, 0, 0, 0]
+
+        # move the core account data to the advice stack
+        adv.push_mapval
+        # OS => [foreign_account_id, 0, 0, 0]
+        # AS => [[foreign_account_id, 0, 0, account_nonce], VAULT_ROOT, STORAGE_ROOT, CODE_ROOT]
+
+        # store the id and nonce of the foreign account to the memory
+        dropw adv_loadw exec.memory::set_acct_id_and_nonce dropw
+        # OS => []
+        # AS => [VAULT_ROOT, STORAGE_ROOT, CODE_ROOT]
+
+        # store the vault root of the foreign account to the memory
+        adv_loadw exec.memory::set_acct_vault_root dropw
+        # OS => []
+        # AS => [STORAGE_ROOT, CODE_ROOT]
+
+        ####
+        # This two chunks of code bellow are essentially doing the same thing, the problem why I 
+        # can't create a dedicated procedure for them is that they're using different memory 
+        # procedures. I can create two new procedures in memory and move the corresponding code 
+        # there, although I'm not sure which approach will be better.
+        ####
```

Comment:
I don't mind these being separate - but a coupe of comments:

- I would probably create two separate procedures - e.g., something like `load_account_storage_data` and `load_account_procedure_data`.
- This code is probably very similar to what we do in the prologue for the native account. Is there a way to use these procedures in both places?
- Why do we use `exec.mem::pipe_words_to_memory` rather than `exec.mem::pipe_double_words_to_memory` here? As far as I know, both storage slots and procedures are take up 2 works.

Reply:
I think the best option will be to move the `validate_account_procedures` and `validate_account_storage_slots` procedures from the prologue to the memory and use them in both places: in prologue and in api. That will solve the highlighted problems.